compiler/aarch64/runtime/ARM64arrayCopy.spp

```diff
@@ -0,0 +1,258 @@
+
+// Copy array elements
+//
+// in:    x0 - length in bytes
+//        x1 - src addr
+//        x2 - dst addr
+// trash: x3, x4
+
+__arrayCopy:
+	cbz	x0, finished			// return if no bytes to copy
+	subs	x3, x2, x1
+	beq	finished			// return if srcAddr == dstAddr
+	cmp	x0, x3
+	bhi	__backwardArrayCopy		// byteLength > dstAddr - srcAddr, must do backward array copy
+	// Forward copy case: fall through
+
+__forwardArrayCopy:
+	tst	x2, #1
+	beq	fwDstAlign2			// dstAddr is 2-byte aligned
+	ldrb	w3, [x1], #1
+	sub	x0, x0, #1
+	strb	w3, [x2], #1
+fwDstAlign2:
+	cmp	x0, #2
+	blt	fwByteCopy			// less than 2 bytes remaining
+	tst	x1, #1
+	bne	fwByteCopyLoop			// srcAddr is not 2-byte aligned
+	tst	x2, #2
+	beq	fwDstAlign4			// dstAddr is 4-byte aligned
+	ldrh	w3, [x1], #2
+	sub	x0, x0, #2
+	strh	w3, [x2], #2
+fwDstAlign4:
+	cmp	x0, #4
+	blt	fwHalfWordCopy			// less than 4 bytes remaining
+	tst	x1, #2
+	bne	fwHalfWordCopyLoop		// srcAddr is not 4-byte aligned
+	tst	x2, #4
+	beq	fwDstAlign8			// dstAddr is 8-byte aligned
+	ldr	w3, [x1], #4
+	sub	x0, x0, #4
+	str	w3, [x2], #4
+fwDstAlign8:
+	cmp	x0, #8
+	blt	fwWordCopy			// less than 8 bytes remaining
+	tst	x1, #4
+	bne	fwWordCopyLoop			// srcAddr is not 8-byte aligned
+	tst	x2, #8
+	beq	fwDstAlign16			// dstAddr is 16-byte aligned
+	ldr	x3, [x1], #8
+	sub	x0, x0, #8
+	str	x3, [x2], #8
+fwDstAlign16:
+	tst	x1, #8
+	bne	fwDoubleWordCopyLoop		// srcAddr is not 16-byte aligned
+fwQuadWordCopyLoop:
+	// Both srcAddr and dstAddr are 16-byte aligned
+	cmp	x0, #16
+	blt	fwDoubleWordCopy		// less than 16 bytes remaining
+	ldp	x3, x4, [x1], #16
+	sub	x0, x0, #16
+	stp	x3, x4, [x2], #16
+	b	fwQuadWordCopyLoop
+fwDoubleWordCopyLoop:
+	// Both srcAddr and dstAddr are 8-byte aligned
+	cmp	x0, #8
+	blt	fwWordCopy			// less than 8 bytes remaining
+	ldr	x3, [x1], #8
+	sub	x0, x0, #8
+	str	x3, [x2], #8
+	b	fwDoubleWordCopyLoop
+fwWordCopyLoop:
+	// Both srcAddr and dstAddr are 4-byte aligned
+	cmp	x0, #4
+	blt	fwHalfWordCopy
+	ldr	w3, [x1], #4			// less than 4 bytes remaining
+	sub	x0, x0, #4
+	str	w3, [x2], #4
+	b	fwWordCopyLoop
+fwHalfWordCopyLoop:
+	// Both srcAddr and dstAddr are 2-byte aligned
+	cmp	x0, #2
+	blt	fwByteCopy			// less than 2 bytes remaining
+	ldrh	w3, [x1], #2
+	sub	x0, x0, #2
+	strh	w3, [x2], #2
+	b	fwHalfWordCopyLoop
+fwByteCopyLoop:
+	cbz	x0, finished
+	ldrb	w3, [x1], #1
+	sub	x0, x0, #1
+	strb	w3, [x2], #1
+	b	fwByteCopyLoop
+fwDoubleWordCopy:
+	// Both srcAddr and dstAddr are 8-byte aligned
+	cmp	x0, #8
+	blt	fwWordCopy
+	ldr	x3, [x1], #8
+	sub	x0, x0, #8
+	str	x3, [x2], #8
+fwWordCopy:
+	// Both srcAddr and dstAddr are 4-byte aligned
+	cmp	x0, #4
+	blt	fwHalfWordCopy
+	ldr	w3, [x1], #4
+	sub	x0, x0, #4
+	str	w3, [x2], #4
+fwHalfWordCopy:
+	// Both srcAddr and dstAddr are 2-byte aligned
+	cmp	x0, #2
+	blt	fwByteCopy
+	ldrh	w3, [x1], #2
+	sub	x0, x0, #2
+	strh	w3, [x2], #2
+fwByteCopy:
+	cbz	x0, finished
+	ldrb	w3, [x1], #1
+	sub	x0, x0, #1
+	strb	w3, [x2], #1
+finished:
+	ret
+
+__backwardArrayCopy:
```

Comment:
If this is an external callable entry point, please provide some documentation on the API and what it accomplishes functionally.

Reply:
Thanks for the comment, but if this is callable from JITed code please describe the API like you do for _arraycopy.  For example (tailor it to what the register usage actually is):
```
// in:    x0 - length in bytes
//        x1 - src addr
//        x2 - dst addr
// trash: x3, x4
```